textworld/generator/chaining.py

```diff
@@ -296,3 +296,211 @@ def print_chains(chains, verbose=False, backward=False):
+
+
+class _ChainNode:
+    """
+    A node in a chain being generated.
+    """
+
+    def __init__(self, parent, state, action, backtracks, depth, breadth):
+        self.parent = parent
+        self.state = state
+        self.action = action
+        self.backtracks = backtracks
+        self.depth = depth
+        self.breadth = breadth
+
+
+class _Chainer:
+    """
+    Helper class for the chaining implementation.
+    """
+
+    def __init__(self, state, backward, min_depth, max_depth, max_breadth,
+                     create_variables, logic, rules_per_depth):
+        self.state = state
+        self.backward = backward
+        self.min_depth = min_depth
+        self.max_depth = max_depth
+        self.max_breadth = max_breadth
+        self.create_variables = create_variables
+
+        if logic is None:
+            self.logic = get_logic()
+        else:
+            self.logic = logic
+
+        self.rules = self.logic.rules.values()
+        self.constraints = self.logic.constraints.values()
+
+        if rules_per_depth is None:
+            self.rules_per_depth = {}
+        else:
+            self.rules_per_depth = rules_per_depth
+
+    def root(self):
+        return _ChainNode(None, self.state, None, [], 0, 1)
+
+    def children(self, node):
+        if node.depth < self.max_depth:
+            yield from self.chain(node)
+
+        yield from self.backtrack(node)
+
+    def chain(self, node):
+        rules = self.rules_per_depth.get(node.depth, self.rules)
+
+        actions = []
+        states = []
+        for rule, mapping in self.all_assignments(node.state, rules):
+            action = self.try_instantiate(node.state, rule, mapping)
+            if not action:
+                continue
+
+            if not self.is_relevant(node, action):
+                continue
+
+            state = self.apply(node.state, action)
+            if not state:
+                continue
+
+            actions.append(action)
+            states.append(state)
+
+        for i, action in enumerate(actions):
+            remaining = actions[i+1:]
+            backtracks = node.backtracks + [remaining]
+            yield _ChainNode(node, states[i], action, backtracks, node.depth + 1, node.breadth)
+
+    def backtrack(self, node):
+        if node.breadth >= self.max_breadth:
+            return
+
+        for i, actions in enumerate(node.backtracks):
+            backtracks = node.backtracks[:i]
+
+            for j, action in enumerate(actions):
+                state = self.apply(node.state, action)
+                if not state:
+                    continue
+
+                remaining = actions[j+1:]
+                new_backtracks = backtracks + [remaining]
+                yield _ChainNode(node, state, action, new_backtracks, i + 1, node.breadth + 1)
+
+    def is_relevant(self, node, action):
+        if not node.action:
+            return True
+
+        if self.backward:
+            # XXX
```

Comment:
What are those? Your TODOs?

Reply:
Yeah, for me a TODO might survive into master and be fixed later, whereas an XXX needs to be fixed ASAP.